web/src/engine/websites/decorators/Grouple.ts

```diff
@@ -0,0 +1,111 @@
+ * @param this - A reference to the {@link MangaScraper}
+ * @param chapter - A reference to the {@link Chapter} which contains the pages
+ * @param script - A script to extract the pages as \{ url : string, mirrors: string[] \}
+ */
+async function FetchPagesSinglePageJS(this: MangaScraper, chapter: Chapter, script: string = pagesWithServersScript, delay: number = 1500): Promise<Page[]> {
+    const uri = new URL(chapter.Identifier, this.URI);
+    uri.searchParams.set('mtr', '1');
+    const images = await FetchWindowScript<ImagesData[]>(new Request(uri), script, delay);
+    return images.map(image => new Page<PageParameters>(this, chapter, new URL(image.url), { mirrors: image.mirrors }));
+}
+/**
+ * A class decorator that adds the ability to extract all pages for a given chapter using the given JS {@link script}.
+ * The pages are extracted from the composed url based on the `Identifier` of the chapter and the `URI` of the website.
+ * @param script - A script to extract the pages as \{ url : string, mirrors: string[] \}
+ * @param delay - An initial delay [ms] before the {@link script} is executed
+ */
+export function PagesSinglePageJS(script = pagesWithServersScript, delay = 0) {
+    return function DecorateClass<T extends Common.Constructor>(ctor: T, context?: ClassDecoratorContext): T {
+        Common.ThrowOnUnsupportedDecoratorContext(context);
+
+        return class extends ctor {
+            public async FetchPages(this: MangaScraper, chapter: Chapter): Promise<Page[]> {
+                return FetchPagesSinglePageJS.call(this, chapter, script, delay);
+            }
+        };
+    };
+}
+
+/**
+ * A class decorator that adds the ability to get the image data for a given page by loading the source asynchronous with the `Fetch API`.
+ * This method fetch pages by trying different urls. Page must have "mirrors: url,url, url, etc..." in Parameters.
+ */
+export function ImageAjaxWithMirrors() {
+    return function DecorateClass<T extends Common.Constructor>(ctor: T, context?: ClassDecoratorContext): T {
+        Common.ThrowOnUnsupportedDecoratorContext(context);
+        return class extends ctor {
+            public async FetchImage(this: MangaScraper, page: Page, priority: Priority, signal: AbortSignal): Promise<Blob> {
+                return FetchMirroredImage.call(this, page, priority, signal);
+            }
+        };
+    };
+}
+
+async function FetchMirroredImage(this: MangaScraper, page: Page<PageParameters>, priority: Priority, signal: AbortSignal): Promise<Blob> {
+    return this.imageTaskPool.Add(async () => {
+        for (const uri of [page.Link, ...page.Parameters.mirrors]) {
+            try {
+                const request = new Request(uri, { signal: signal, headers: { Referer: this.URI.href } });
+                const response = await Fetch(request);
+                const blob = await response.blob();
+                if (blob.type.startsWith('image/')) return blob;
+            } catch { }
+        }
```

Comment:
Should there be a specific error in case all requests have failed or is returning an `undefined` blob sufficient?

Reply:
Wont the calling code throw an exception anyway ?

Reply:
This depends on the caller and how the returned `undefined` is processed.
The question still stands, is this behavior sufficient or should there maybe an explicit (accumulated) error for this?

Reply:
![image](https://github.com/user-attachments/assets/2845a275-a52d-43a2-939e-5d55694980a5)